AEMG/data_utils.py

```diff
@@ -95,12 +95,54 @@ def __init__(self,config):
         self.labels = np.array(self.labels)
         self.labels = torch.from_numpy(self.labels).long()
 
+        self.generate_opposite_pairs()
+
     def __len__(self):
-        return len(self.final_points)
+        return len(self.opposite_pairs)
 
     def __getitem__(self, idx):
-        return self.final_points[idx], self.labels[idx]         
-            
+        return self.opposite_pairs[idx]
```

Comment:
Can't we just return `self.final_points[self.opposite_pairs[idx][0],  self.final_points[self.opposite_pairs[idx][1]]` and remove the need for a `collate_fn`?

Reply:
We will still need the collate function to gather the relevant final points in that batch and change the indices correspondingly

Reply:
Here is my understanding:
* For `S` successful and `F` failure points in the dataset, `self.opposite_pairs` is an `(SxF, 2)` shaped array where an index contains a success and a failure.
* The default implementation of `collate_fn` just contains multiple calls to `__getitem__` (see the implementation for `DynamicsDataset`, which doesn't use a `collate_fn`.
* So, if the implementation of `__getitem__` already returns a tuple of `self.final_points`, where the first element of each tuple is a success and the second element is a failure, we do not need a `collate_fn`.

Lmk if this makes sense.

Reply:
It does make sense. But I'm not sure if it is efficient. Let's say, we have a success point `s1` and we have `F` failure points. Then `s1` is present `F` times in `self.opposite_pairs`. Now, if in the current batch `s1` appears more than once, then by sending `self.final_points[s1]`  multiple times, we will have to encode it multiple times. Instead by using the collate function, we are reducing the number of encodings. So I still feel using the collate function is more efficient

Reply:
Ah, good catch! OK, then make sure it's getting passed as an argument to the loader and we should be good.